### PR TITLE
feat: new swap API implementation

### PR DESCRIPTION
**Pull request by pacmanifold**
> [!IMPORTANT]
> Introduces a new swap API for the DEX, adding exact input/output swap functions, updating models and queries, and providing extensive tests.
> 
>   - **Swap API Implementation**:
>     - Adds `swap_exact_amount_in` and `swap_exact_amount_out` functions in `liquidity_pool.rs` and `router.rs` for handling swaps with exact input and output amounts.
>     - Updates `execute.rs` to handle new swap messages `SwapExactAmountIn` and `SwapExactAmountOut`.
>     - Implements query support for swap simulations in `query.rs`.
>   - **Models and Types**:
>     - Introduces `SwapRoute` type in `msgs.rs` to define swap paths with a maximum length of 2.
>     - Adds `PairId` struct in `msgs.rs` for identifying trading pairs.
>     - Updates `events.rs` with new events `SwapExactAmountIn` and `SwapExactAmountOut`.
>   - **Testing**:
>     - Extensive tests added in `dex.rs` for swap functionality, including edge cases and error handling.
>     - Updates `setup.rs` to include necessary configurations for testing swaps.
>   - **Miscellaneous**:
>     - Minor refactoring in `core.rs` and `coin_pair.rs` for improved code clarity and efficiency.
>     - Adjusts `Bound` implementations in `bound.rs` for better handling of bounded values.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=left-curve%2Fleft-curve&utm_source=github&utm_medium=referral)<sup> for d1fee309fe4d68a76bb916a283f9c2e0b7cde3b6. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: dango/dex/src/execute.rs [hunk 1]
```diff
@@ -348,6 +357,74 @@ fn withdraw_liquidity(
...
+    }
+
+    // Save the updated pool reserves.
+    for (pair, reserve) in reserves {
+        RESERVES.save(ctx.storage, (&pair.base_denom, &pair.quote_denom), &reserve)?;
+    }
+
+    // TODO:
+    // 1. add events
+    // 2. handle the case if output is zero
```
Comment (pacmanifold) [inline dango/dex/src/execute.rs:386]: should we just set `minimum_out = minimum_out.unwrap_or(Uint128::ZERO)?`? then the tx will fail if output is zero. seems most reasonable to me

File: dango/dex/src/core/liquidity_pool.rs [hunk 2]
```diff
@@ -47,6 +51,77 @@ pub trait PassiveLiquidityPool {
...
+    /// - `output`: The amount of output asset to receive.
+    ///
+    /// ## Outputs
+    ///
+    /// - The amount of input asset required to receive the exact output.
+    fn simulate_swap_exact_amount_out(
+        &self,
+        reserve: CoinPair,
+        output: Coin,
+    ) -> anyhow::Result<Coin>;
```
[the remark this answers is not in the document]
Comment (pacmanifold) [reply dango/dex/src/core/liquidity_pool.rs:124]: yeah just realised they are unnecessary. removing now

File: dango/dex/src/execute.rs (hunk 1 above)
Comment (pacmanifold) [reply dango/dex/src/execute.rs:386]: @larry0x I implemented it like this now

**Comment by larry0x** [comment]
@ellipsis can you review this please

File: dango/testing/src/setup.rs [hunk 3]
```diff
@@ -241,6 +241,8 @@ where
                 balances: coins! {
                     DANGO_DENOM.clone() => 100_000_000_000_000,
                     USDC_DENOM.clone()  => 100_000_000_000,
+                    ETH_DENOM.clone() => 100_000_000_000_000,
+                    BTC_DENOM.clone() => 100_000_000_000_000,
```
Comment (larry0x) [inline dango/testing/src/setup.rs:245]: Can you align the `=>` symbols

File: dango/testing/src/setup.rs [hunk 4]
```diff
@@ -253,6 +255,7 @@ where
                     // synth tokens in genesis. We add this just for testing purpose.
                     WBTC_DENOM.clone() => 100_000_000_000_000,
                     ETH_DENOM.clone()  => 100_000_000_000_000,
+                    BTC_DENOM.clone() => 100_000_000_000_000,
```
Comment (larry0x) [inline dango/testing/src/setup.rs:258]: Align the `=>` here as well

File: dango/types/src/dex/events.rs [hunk 5]
```diff
@@ -1,4 +1,5 @@
 use {
+    super::PairId,
     crate::dex::{Direction, OrderId},
```
Comment (larry0x) [inline dango/types/src/dex/events.rs:3]: I don't like having `super` and `crate` in the same `use` statement

```suggestion
    crate::dex::{Direction, OrderId, PairId},
```

File: grug/types/src/coin_pair.rs [hunk 6]
```diff
@@ -160,6 +160,63 @@ impl CoinPair {
...
+    pub fn amount_of(&self, denom: &Denom) -> StdResult<Uint128> {
+        if self.first().denom == denom {
+            Ok(*self.first().amount)
+        } else if self.second().denom == denom {
+            Ok(*self.second().amount)
+        } else {
+            Err(StdError::invalid_coins(format!(
+                "coin pair {:?} doesn't have denom: {}",
+                self, denom
+            )))
```
Comment (larry0x) [inline grug/types/src/coin_pair.rs:179]: Change

```rust
format!("coin pair {:?} doesn't have denom: {}", self, denom)
```

to

```rust
format!("coin pair {self:?} doesn't have denom: {denom}")
```

Same with the two other `format!` calls below.

File: dango/dex/src/core/liquidity_pool.rs [hunk 7]
```diff
@@ -114,6 +160,94 @@ impl PassiveLiquidityPool for PairParams {
...
+        let a = reserve.amount_of(&input.denom)?;
+        let b = reserve.amount_of(&output_denom)?;
+        let out_amount = match self.curve_invariant {
+            CurveInvariant::Xyk => {
+                // Solve A * B = (A + offer.amount) * (B - amount_out) for amount_out
+                // => amount_out = B - (A * B) / (A + offer.amount)
+                // Round so that user takes the loss
+                let amount_out = b.checked_sub(
+                    Uint128::ONE.checked_multiply_ratio_ceil(a * b, a + input.amount)?,
+                )?;
```
Comment (larry0x) [inline dango/dex/src/core/liquidity_pool.rs:186]: 1. It's better we use checked math instead `*` and `-`
2. Why not this:

```diff
  let amount_out = b.checked_sub(
-    Uint128::ONE.checked_multiply_ratio_ceil(a * b, a + input.amount)?,
+    a.checked_multiply_ratio(b, a.checked_add(input.amount)?)?,
  )?;
```

Comment (larry0x) [reply dango/dex/src/core/liquidity_pool.rs:186]: I made this change and all tests pass. I think they are equivalent.

File: dango/dex/src/core/liquidity_pool.rs [hunk 8]
```diff
@@ -114,6 +160,94 @@ impl PassiveLiquidityPool for PairParams {
...
+        let offer_reserves = reserve.amount_of(&denom_in)?;
+        let ask_reserves = reserve.amount_of(&output.denom)?;
+        ensure!(ask_reserves > output.amount, "insufficient liquidity");
+
+        let amount_in = match self.curve_invariant {
+            CurveInvariant::Xyk => {
+                // Apply swap fee. In SwapExactIn we multiply ask by (1 - fee) to get the
+                // offer amount after fees. So in this case we need to divide ask by (1 - fee)
+                // to get the ask amount after fees. Round so that user takes the loss
+                let coin_out_after_fee = output
```
Comment (larry0x) [inline dango/dex/src/core/liquidity_pool.rs:226]: This is more appropriately called `before_fee` imo

File: dango/dex/src/core/router.rs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+        // Save the updated pool reserve.
+        reserves.insert(pair.clone(), reserve);
+
+        // Add events
+        events.push(SwapExactAmountIn {
+            pair,
+            input,
+            output: output.clone(),
+        })?;
```
Comment (larry0x) [inline dango/dex/src/core/router.rs:40]: I feel we should have only one event regardless of how many step there are in the route. 

The purpose of events if for notifications on the frontend.

If I swap A -> B -> C, I want to only receive one notification "you swapped A for C", not two notifications "you swapped A for B" and "you swapped B for C",

File: dango/dex/src/execute.rs [hunk 10]
```diff
@@ -348,6 +357,73 @@ fn withdraw_liquidity(
...
+
+    // Ensure the input is below the maximum.
+    if let Some(maximum_input) = maximum_input {
+        ensure!(
+            input.amount <= maximum_input,
+            "input amount is above the maximum: {} > {}",
+            input.amount,
+            maximum_input
+        );
+    }
```
[the remark this answers is not in the document]
Comment (pacmanifold) [reply dango/dex/src/execute.rs]: true

File: dango/dex/src/core/liquidity_pool.rs (hunk 7 above)
Comment (pacmanifold) [reply dango/dex/src/core/liquidity_pool.rs]: yes should be equivalent

Comment (pacmanifold) [reply dango/dex/src/core/liquidity_pool.rs]: that's definitely better. don't know why I didn't use checked math in that instance

File: grug/types/src/coin_pair.rs (hunk 6 above)
Comment (pacmanifold) [reply grug/types/src/coin_pair.rs]: fixed
